// metalminer/index.js
(function () {
    'use strict';
    var request = require('request'),
        cheerio = require('cheerio');

    var mm = module.exports = {};

    mm.options = {
        timeout: 2000
    };

    mm.getBandInfo = function (metaInfo, callback) {
        var functions = [
            {
                name: 'Wikipedia',
                func: getWikipediaBandInfo,
                prio: 1,
                status: 'NOT_STARTED',
                data: '',
                error: ''
            }
        ];
        var callback2 = function (err, res) {
            var currObj = {};
            for (var i = 0; i < functions.length; i++) {
                if (res.caller.name === functions[i].name) {
                    currObj = functions[i];
                }
            }
            currObj.status = 'DONE';
            currObj.data = res.data;
            currObj.error = res.error;
            var statuses = functions.map(function (single) {
                return single.status;
            });
            if (statuses.length === statuses.filter(function (single) { return single === 'DONE'; }).length) {
                // All done
                var results = functions.filter(function (single) {
                    return !!single.data;
                }).sort(function (a, b) {
                    return (a.prio > b.prio);
                });
                if (results.length === 0) {
                    callback('Not found');
                } else {
                    callback(null, results[0].data);
                }
            }
        };

        for (var i = 0, endi = functions.length; i < endi; i++) {
            var currFunc = functions[i];
            currFunc.status = 'SEARCHING';
            currFunc.func.call(currFunc, metaInfo, callback2);
        }
    }

    mm.getSimilarArtists = function (metaInfo, callback) {
        var functions = [
            {
                name: 'Metal Archives',
                func: getMetalArchivesSimilarArtist,
                prio: 1,
                status: 'NOT_STARTED',
                bands: '',
                error: ''
            }
        ];

        var callback2 = function (err, res) {
            var currObj = {};
            for (var i = 0; i < functions.length; i++) {
                if (res.caller.name === functions[i].name) {
                    currObj = functions[i];
                }
            }
            currObj.status = 'DONE';
            currObj.bands = res.bands;
            currObj.error = res.error;
            var statuses = functions.map(function (single) {
                return single.status;
            });
            if (statuses.length === statuses.filter(function (single) { return single === 'DONE'; }).length) {
                // All done
                var results = functions.filter(function (single) {
                    return !!single.bands;
                }).sort(function (a, b) {
                    return (a.prio > b.prio);
                });
                if (results.length === 0) {
                    callback('Not found');
                } else {
                    callback(null, results[0].bands);
                }
            }
        };

        for (var i = 0, endi = functions.length; i < endi; i++) {
            var currFunc = functions[i];
            currFunc.status = 'SEARCHING';
            currFunc.func.call(currFunc, metaInfo, callback2);
        }
    };

    mm.getLyrics = function (metaInfo, callback) {
        var functions = [
            {
                name: 'Metal Archives',
                func: getMetalArchivesLyrics,
                prio: 1,
                status: 'NOT_STARTED',
                lyrics: '',
                error: ''
            },
            {
                name: 'Dark Lyrics',
                func: getDarkLyrics,
                prio: 2,
                status: 'NOT_STARTED',
                lyrics: '',
                error: ''
            },
            {
                name: 'Song Lyrics',
                func: getSongLyrics,
                prio: 3,
                status: 'NOT_STARTED',
                lyrics: '',
                error: ''
            }
        ];

        var callback2 = function (err, res) {
            var currObj = {};
            for (var i = 0; i < functions.length; i++) {
                if (res.caller.name === functions[i].name) {
                    currObj = functions[i];
                }
            }
            currObj.status = 'DONE';
            currObj.lyrics = res.lyrics;
            currObj.error = res.error;
            var statuses = functions.map(function (single) {
                return single.status;
            });
            if (statuses.length === statuses.filter(function (single) { return single === 'DONE'; }).length) {
                // All done
                var results = functions.filter(function (single) {
                    return !!single.lyrics;
                }).sort(function (a, b) {
                    return (a.prio > b.prio);
                });
                if (results.length === 0) {
                    callback('Not found');
                } else {
                    callback(null, results[0].lyrics);
                }
            }
        };

        for (var i = 0, endi = functions.length; i < endi; i++) {
            var currFunc = functions[i];
            currFunc.status = 'SEARCHING';
            currFunc.func.call(currFunc, metaInfo, callback2);
        }
    };

    var getWikipediaBandInfo = function (metaInfo, callback) {
        if (!metaInfo || !callback) {
            return;
        }
        var that = this;
        var uri = 'http://www.google.com/search?q=' + metaInfo.artist.replace(/ /g, '+') + '+band+site:en.wikipedia.org';
        uri = encodeURI(uri);
        request({ uri: uri, headers: { 'User-Agent': 'metalminer/0.1 (http://jooon.mooo.com:8088; jon.asplund@gmail.com)'} }, function (err, res, body) {
            if (err) {
                callback('Unknown error', { caller: that });
                return;
            }
            var $ = cheerio.load(body, { ignoreWhitespace: true });
            var link = $('body h3 a')[0];
            if (!link || !link.attribs || !link.attribs['href']) {
                callback('Malformed response from google', { caller: that });
                return;
            }
            var linkhref = link.attribs['href'];
            linkhref = linkhref.replace('en.wikipedia.org', 'en.mobile.wikipedia.org').replace('/url?q=', '').replace(/&amp;sa=.*/, '');
            request({ uri: linkhref, headers: { 'User-Agent': 'metalminer/0.1 (http://jooon.mooo.com:8088; jon.asplund@gmail.com)'} }, function (err, res, body) {
                if (err) {
                    console.log(err);
                    callback('Unknown error', { caller: that });
                    return;
                }
                var $ = cheerio.load(body, { ignoreWhitespace: true });
                if ($('h1').text().toLowerCase().indexOf(metaInfo.artist.toLowerCase()) < 0) {
                    callback('Band does not have a wikipedia entry', { caller: that });
                    return;
                }
                var content = $('#content');
                $('.section').each(function (i, item) {
                    var itemText = $(item).text().toString();
                    if (itemText &&
                        itemText.match(/^[References|Read in another language|External links|Bibliography]/i)) {
                        $(item).remove();
                    }
                });
                //$('.metadata').remove();
                var html = content.html().replace(/Jump back a section/g, '');
                callback(null, { caller: that, data: html });
            });
        });
    }

    var getDarkLyrics = function (metaInfo, callback) {
        if (!metaInfo || !callback) {
            return;
        }
        var that = this;
        var uri = 'http://www.darklyrics.com/lyrics/';
        var artist = metaInfo.artist.toLowerCase().replace(/[^a-z]/gi, '');
        var album = metaInfo.album.toLowerCase().replace(/[^a-z]/gi, '');
        var title = metaInfo.title.toLowerCase().replace(/[^a-z]/gi, '');
        uri += artist + '/' + album + '.html';
        request(uri, function (err, res, body) {
            var $ = cheerio.load(body, { ignoreWhitespace: true }),
                $albumlyrics = $('.lyrics'),
                albumlyrics = $albumlyrics.html(),
                $songtitles = $albumlyrics.find('h3'),
                songnr = 0;
            $songtitles.each(function (key) {
                if (this.text().toLowerCase().replace(/[^a-z]/gi, '') === title) {
                    songnr = key;
                    return false;
                }
            });
            if (songnr === 0) {
                callback('Not found', { caller: that });
            } else {
                var songlyrics = albumlyrics.split(/<h3>\s*<a name=".*">.*<\/h3>/);
                if (songlyrics.length < songnr + 1) {
                    callback('Unknown error', { caller: that });
                } else {
                    callback(null, { caller: that, lyrics: songlyrics[songnr + 1] });
                }
            }
        });
    };

    var getMetalArchivesLyrics = function (metaInfo, callback) {
        if (!metaInfo || !callback) {
            return;
        }
        var that = this;
        var uri = 'http://www.metal-archives.com/search/ajax-advanced/searching/songs/?';
        uri += 'songTitle=' + metaInfo.title.replace(/^[0-9]{2,3} - /, '') + '&';
        uri += 'bandName=' + metaInfo.artist + '&';
        uri += 'releaseTitle=' + metaInfo.album;
        uri = uri.replace(/ /g, '+');
        request(uri, function (err, res, body) {
            if (err) {
                callback('Unknown error', { caller: that });
                return;
            }
            try {
                body = JSON.parse(body);
            } catch (e) {
                callback('Did not get JSON response.', { caller: that });
                return;
            }
            if (body.error !== '' || parseInt(body.iTotalDisplayRecords, 10) < 1) {
                callback('Not found.', { caller: that });
                return;
            }
            if (body.aaData.length > 0 && body.aaData[0].length > 4) {
                var link = body.aaData[0][4];
                var match = link.match(/id="lyricsLink_(.*?)"/)[1];
                var href = 'http://www.metal-archives.com/release/ajax-view-lyrics/id/' + match;
                request(href, function (err, res, body2) {
                    if (err) { callback('Unknown error.', { caller: that }); }
                    if (typeof callback === 'function') {
                        callback(null, { caller: that, lyrics: body2 });
                    }
                });
            }
        });
    };

    var getSongLyrics = function (metaInfo, callback) {
        if (!metaInfo || !callback) {
            return;
        }
        var that = this;
        var uri = 'http://www.songlyrics.com/' + metaInfo.artist.toLowerCase().replace(/ /gi, '-') + '/';
        uri += metaInfo.title.toLowerCase().replace(/^[0-9]{2,3} - /, '').replace(/ /gi, '-') + '-lyrics';
        request(uri, function (err, res, body) {
            if (err) {
                callback('Unknown error', { caller: that });
                return;
            }
            var $ = cheerio.load(body, { ignoreWhitepace: true });
            $('script').remove();
            if ($('#songLyricsDiv').length === 0) {
                callback('Song not found.', { caller: that });
            } else {
                if ($('#songLyricsDiv').children().length > 0 &&
                    $('#songLyricsDiv').text().indexOf('Sorry, we have no') < 0) {
                    callback(null, { caller: that, lyrics: $('#songLyricsDiv').html() });
                } else {
                    callback('Could not find lyrics.', { caller: that });
                }
            }
        });
    };

    var getMetalArchivesSimilarArtist = function (metaInfo, callback) {
        if (!metaInfo || !callback) {
            return;
        }
        var that = this;
        var uri = 'http://www.metal-archives.com/search/ajax-advanced/searching/albums/?bandName=';
        uri += metaInfo.artist.replace(/ /gi, '+') + '&releaseTitle=' + metaInfo.album.replace(/ /gi, '+');
        request.get(uri, function (err, res, body) {
            if (err) {
                callback('Unknown error', { caller: that });
                return;
            }
            try {
                body = JSON.parse(body);
            } catch (e) {
                callback('Did not get JSON response.', { caller: that });
            }
            //console.log(body);
            if (body.error !== '' || parseInt(body.iTotalDisplayRecords, 10) < 1) {
                callback('Not found.', { caller: that });
                return;
            }
            if (body.aaData.length > 0 && body.aaData[0].length > 0) {
                var link = body.aaData[0][0];
                var match = link.match(/(?:\/bands\/.*?\/)(.*?)(?:">)/)[1];
                var uri2 = 'http://www.metal-archives.com/band/ajax-recommendations/id/';
                uri2 += match;
                request.get(uri2, function (err, res, body) {
                    if (err) {
                        callback('Unknown error', { caller: that });
                        return;
                    }
                    var match, regex = /(?:\/bands\/)([^\/]*)(?:\/)/gi;
                    var bands = [];
                    while ((match = regex.exec(body)) !== null) {
                        bands.push(match[1].replace(/_/gi, ' '));
                    }
                    if (bands.length > 0) {
                        callback(null, { caller: that, bands: bands });
                    } else {
                        callback('No bands found.', { caller: that });
                    }
                });
            }
        });
    };
})();
