// metalminer/testindex.js
(function () {
    'use strict';
    var request = require('request'),
        cheerio = require('cheerio');
    var mm = module.exports = {};

    var FunctionCaller = function (functions, metaInfo) {
        this.functions = [];
        for (var i = 0, endi = functions.length; i < endi; i++) {
            this.functions[i] = {
                name: functions[i].name || 'function' + i,
                func: functions[i].func,
                prio: functions[i].prio || 1,
                status: 'NOT_STARTED',
                data: ''
            };
        }
        this.metaInfo = metaInfo;
    };

    FunctionCaller.prototype.run = function (callback) {
        var cb = function (err, data) {
            console.log('cb');
            console.log(this);
            console.log('cb');
        };
        for (var i = 0, endi = this.functions.length; i < endi; i++) {
            var currFunc = this.functions[i];
            currFunc.func.call(this, this.metaInfo, cb);
        }
    };

    FunctionCaller.prototype.checkResults = function () {
        console.log(this);
        this.status = 'DONE';
        currFunc.data = data;
        if (err) {
            callback(err);
        } else {
            callback(null, data);
        }
    }

    mm.getLyrics = function (metaInfo, callback) {
        var functions = [
            {
                func: getDarkLyrics,
                prio: 2
            }
        ];
        var fc = new FunctionCaller(functions, metaInfo);
        fc.run(callback);
    };

    FunctionCaller.prototype.getDarkLyrics = function (metaInfo, callback) {
        if (!metaInfo || !callback) {
            return;
        }
        var uri = 'http://www.darklyrics.com/lyrics/';
        var artist = metaInfo.artist.toLowerCase().replace(/[^a-z]/gi, '');
        var album = metaInfo.album.toLowerCase().replace(/[^a-z]/gi, '');
        var title = metaInfo.title.toLowerCase().replace(/[^a-z]/gi, '');
        uri += artist + '/' + album + '.html';
        request(uri, function (err, res, body) {
            var $ = cheerio.load(body, { ignoreWhitespace: true }),
                $albumlyrics = $('.lyrics'),
                albumlyrics = $albumlyrics.html(),
                $songtitles = $albumlyrics.find('h3'),
                songnr = 0;
            $songtitles.each(function (key) {
                if (this.text().toLowerCase().replace(/[^a-z]/gi, '') === title) {
                    songnr = key;
                    return false;
                }
            });
            if (songnr === 0) {
                console.log('not found');
                callback('Not found');
            } else {
                var songlyrics = albumlyrics.split(/<h3>\s*<a name=".*">.*<\/h3>/);
                if (songlyrics.length < songnr + 1) {
                    console.log('error');
                    callback('Unknown error');
                } else {
                    console.log('success');
                    callback(null, songlyrics[songnr + 1]);
                }
            }
        });
    };
})();
